subprojects/core/src/test/groovy/org/gradle/api/internal/tasks/DefaultTaskOutputsTest.groovy

```diff
@@ -241,53 +241,96 @@ class DefaultTaskOutputsTest extends Specification {
     }
 
     def "can turn caching on via cacheIf()"() {
+        outputs.dir("someDir")
+
         expect:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
 
         when:
-        outputs.cacheIf { true}
+        outputs.cacheIf { true }
         then:
-        outputs.cacheEnabled
+        outputs.taskCaching.cacheable
     }
 
     def "can turn caching off via cacheIf()"() {
+        outputs.dir("someDir")
+
         expect:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
 
         when:
         outputs.cacheIf { true }
         then:
-        outputs.cacheEnabled
+        outputs.taskCaching.cacheable
 
         when:
         outputs.cacheIf { false }
         then:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
 
         when:
         outputs.cacheIf { true }
         then:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
     }
 
     def "can turn caching off via doNotCacheIf()"() {
+        outputs.dir("someDir")
+
         expect:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
 
         when:
         outputs.doNotCacheIf { false }
         then:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
 
         when:
         outputs.cacheIf { true }
         then:
-        outputs.cacheEnabled
+        outputs.taskCaching.cacheable
 
         when:
         outputs.doNotCacheIf { true }
         then:
-        !outputs.cacheEnabled
+        !outputs.taskCaching.cacheable
+    }
+
+    def "first reason for not caching is reported"() {
+        expect:
+        !outputs.taskCaching.cacheable
+        outputs.taskCaching.disabledReason == "Caching has not been enabled for the task"
+
+        when:
+        outputs.cacheIf { true }
+
+        then:
+        !outputs.taskCaching.cacheable
+        outputs.taskCaching.disabledReason == "No outputs declared"
+
+        when:
+        outputs.dir("someDir")
+
+        then:
+        outputs.taskCaching.cacheable
+
+        when:
+        outputs.doNotCacheIf("Caching manually disabled") { true }
+
+        then:
+        !outputs.taskCaching.cacheable
+        outputs.taskCaching.disabledReason == "Caching manually disabled"
+    }
+
+    def "disabling caching for plural file outputs is reported"() {
+        when:
+        outputs.cacheIf { true }
+        outputs.files("someFile", "someOtherFile")
+
+        then:
+        !outputs.taskCaching.cacheable
+        outputs.taskCaching.disabledReason == "Declares multiple output files for a single output property via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`"
+
```

Comment:
Should we have a test for the value of `disabledReason` when the task _is_ cacheable?